# El módulo os

Si bien este módulo contiene muchos métodos relacionados con el manejo del sistema, se mostrarán los más relevantes para nuestro curso

In [1]:
import os  # importar el módulo

In [2]:
# la herramienta más importante son las funciones del modulo path dentro de os, 
# permite definir direcciones de archivos y directorios, verificar su existencia, etc
print(os.path.abspath('.')) # para obtener el path absoluto de un directorio
print(os.path.isdir('../Tema-1/periodica.png')) # para saber si un elemento es un directorio
print(os.path.join('..','Tema-1')) # para concatenar paths

/home/sebastian/Documents/Codes/series-temporales-machine-learning/Apuntes/Tema-2
False
../Tema-1


In [3]:
# se puede saber la carpeta donde estamos trabajando 
os.getcwd()

'/home/sebastian/Documents/Codes/series-temporales-machine-learning/Apuntes/Tema-2'

In [4]:
# se puede ver el contenido de la carpeta, devuelve una lista
os.listdir()

['Tema-2.1-tipos-de-datos.md',
 'Tema-2.2-funciones-clases-objetos.md',
 '2.2-funciones-clases-objetos.ipynb',
 'Tema-2.4-librerias.md',
 'new_dir',
 'Tema-2.3-manejo-archivos.md',
 '2.1-tipos-de-datos.ipynb',
 'Tema-2-generalidades.md',
 '2.4.1-librerias-estandar.ipynb',
 'python.png']

In [7]:
# se puede crear un directorio
os.mkdir('./new_dir')
print(os.listdir())
# se puede cambiar el directorio de trabajo
os.chdir('./new_dir/')
print(os.listdir())
# se puede volver atras
os.chdir('..')

['Tema-2.1-tipos-de-datos.md', 'Tema-2.2-funciones-clases-objetos.md', '2.2-funciones-clases-objetos.ipynb', 'Tema-2.4-librerias.md', 'new_dir', 'Tema-2.3-manejo-archivos.md', '2.1-tipos-de-datos.ipynb', 'Tema-2-generalidades.md', '2.4.1-librerias-estandar.ipynb', 'python.png']
[]


In [8]:
# se puede eliminar la carpeta
os.rmdir('./new_dir')
os.listdir('.')

['Tema-2.1-tipos-de-datos.md',
 'Tema-2.2-funciones-clases-objetos.md',
 '2.2-funciones-clases-objetos.ipynb',
 'Tema-2.4-librerias.md',
 'Tema-2.3-manejo-archivos.md',
 '2.1-tipos-de-datos.ipynb',
 'Tema-2-generalidades.md',
 '2.4.1-librerias-estandar.ipynb',
 'python.png']

In [13]:
# Supongamos que queremos leer uno de los archivos de la carpeta googleStock
data_folder = os.path.join('..','..','Data','googleStock') #creamos el path

#verificamos que sea correcto y en caso que asi sea, vemos los nombres de archivos dentro
if os.path.isdir(data_folder):
    googleFiles = os.listdir(data_folder) # guardamos la lista
    print(googleFiles)


['Google_Stock_Test (2023).csv', 'Google_Stock_Train (2010-2022).csv']


In [16]:
# para leer el contenido de alguno de ellos podemos usar la lista
with open(os.path.join(data_folder,googleFiles[0]), 'r') as fo:
    data = fo.readlines()

# Imprimimos bonito
for d in data:
    print(d)

Date,Open,High,Low,Close,Adj Close,Volume

2023-01-03,89.589996,91.050003,88.519997,89.120003,89.120003,28131200

2023-01-04,90.349998,90.650002,87.269997,88.080002,88.080002,34854800

2023-01-05,87.470001,87.570000,85.900002,86.199997,86.199997,27194400

2023-01-06,86.790001,87.690002,84.860001,87.339996,87.339996,41381500

2023-01-09,88.360001,90.050003,87.860001,88.019997,88.019997,29003900

2023-01-10,85.980003,88.669998,85.830002,88.419998,88.419998,30467800

2023-01-11,89.180000,91.599998,89.010002,91.519997,91.519997,26862000

2023-01-12,91.480003,91.870003,89.750000,91.129997,91.129997,30258100

2023-01-13,90.849998,92.190002,90.129997,92.120003,92.120003,26329200

2023-01-17,92.059998,92.250000,90.050003,91.290001,91.290001,32602400

2023-01-18,92.139999,92.800003,90.639999,91.120003,91.120003,29116700

2023-01-19,90.720001,93.610001,90.629997,93.050003,93.050003,37000400

2023-01-20,95.099998,98.300003,95.019997,98.019997,98.019997,63191100

2023-01-23,97.949997,100.040001,97

# El módulo csv

Este módulo contiene muchas funciones útiles para leer archivos de tipo csv (comma separated values), incluso tablas de datos con otros separadores

In [18]:
import csv #importamos el módulo

In [23]:
# leyendo un archivo separado por comas con csv.reader()

with open(os.path.join(data_folder, googleFiles[0]), 'r') as fp: # handler del archivo
    print(f'leyendo el archivo "{googleFiles[0]}"...')
    reader = csv.reader(fp, delimiter=',') # creamos el objeto que leera el archivo linea por linea
    data = [] # guardaremos cada linea en una lista
    for row in reader: # recorremos cada linea
        data.append(row) # agregamos cada linea

for r in data: # imprimimos (observar que el resultado es una lista de listas)
    print(r)

leyendo el archivo "Google_Stock_Test (2023).csv"...
['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']
['2023-01-03', '89.589996', '91.050003', '88.519997', '89.120003', '89.120003', '28131200']
['2023-01-04', '90.349998', '90.650002', '87.269997', '88.080002', '88.080002', '34854800']
['2023-01-05', '87.470001', '87.570000', '85.900002', '86.199997', '86.199997', '27194400']
['2023-01-06', '86.790001', '87.690002', '84.860001', '87.339996', '87.339996', '41381500']
['2023-01-09', '88.360001', '90.050003', '87.860001', '88.019997', '88.019997', '29003900']
['2023-01-10', '85.980003', '88.669998', '85.830002', '88.419998', '88.419998', '30467800']
['2023-01-11', '89.180000', '91.599998', '89.010002', '91.519997', '91.519997', '26862000']
['2023-01-12', '91.480003', '91.870003', '89.750000', '91.129997', '91.129997', '30258100']
['2023-01-13', '90.849998', '92.190002', '90.129997', '92.120003', '92.120003', '26329200']
['2023-01-17', '92.059998', '92.250000', '90.050003', '

In [24]:
# para verificar que tenemos una lista de listas podemos imprimir algunos elementos
print(data[0])
print(data[0][1]) # debe ser 'open'
print(data[10])
print(data[10][1]) # debe ser 'open'


['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']
Open
['2023-01-17', '92.059998', '92.250000', '90.050003', '91.290001', '91.290001', '32602400']
92.059998


Se ha visto en el caso anterior que los nombres de las columnas se encuentran en la primera fila. Esto es así porque el archivo fue creado de ese modo. Luego, cada uno de los datos se localiza por medio de la columna (la columna 0 siempre es la fecha, la columna 1 corresponde al precio de apertura, etc). Es posible leer de otra manera los datos de modo que cada uno de ellos pueda ser reconocido por su nombre. Esto se puede hacer mediante diccionarios, donde cada fila corresponde a un diccionario con el nombre de la columna como _key_.

Para esto existe el método DictReader() que se muestra a continuación

In [26]:
# leyendo un archivo separado por comas con csv.DictReader()

with open(os.path.join(data_folder, googleFiles[0]), 'r') as fp: # handler del archivo
    print(f'leyendo el archivo "{googleFiles[0]}"...')
    reader = csv.DictReader(fp, delimiter=',') # creamos el objeto que leera el archivo linea por linea
    data = [] # guardaremos cada linea en una lista
    for row in reader: # recorremos cada linea
        data.append(row) # agregamos cada linea

for r in data: # imprimimos (observar que el resultado es una lista de diccionarios)
    print(r)

leyendo el archivo "Google_Stock_Test (2023).csv"...
{'Date': '2023-01-03', 'Open': '89.589996', 'High': '91.050003', 'Low': '88.519997', 'Close': '89.120003', 'Adj Close': '89.120003', 'Volume': '28131200'}
{'Date': '2023-01-04', 'Open': '90.349998', 'High': '90.650002', 'Low': '87.269997', 'Close': '88.080002', 'Adj Close': '88.080002', 'Volume': '34854800'}
{'Date': '2023-01-05', 'Open': '87.470001', 'High': '87.570000', 'Low': '85.900002', 'Close': '86.199997', 'Adj Close': '86.199997', 'Volume': '27194400'}
{'Date': '2023-01-06', 'Open': '86.790001', 'High': '87.690002', 'Low': '84.860001', 'Close': '87.339996', 'Adj Close': '87.339996', 'Volume': '41381500'}
{'Date': '2023-01-09', 'Open': '88.360001', 'High': '90.050003', 'Low': '87.860001', 'Close': '88.019997', 'Adj Close': '88.019997', 'Volume': '29003900'}
{'Date': '2023-01-10', 'Open': '85.980003', 'High': '88.669998', 'Low': '85.830002', 'Close': '88.419998', 'Adj Close': '88.419998', 'Volume': '30467800'}
{'Date': '2023-01

In [28]:
# veamos como acceder a algunos elementos
print(data[0]) # la primera fila es un diccionario y tiene datos
print(data[0]['Open']) # los datos se localizan a partir de las keys

{'Date': '2023-01-03', 'Open': '89.589996', 'High': '91.050003', 'Low': '88.519997', 'Close': '89.120003', 'Adj Close': '89.120003', 'Volume': '28131200'}
89.589996


Para guardar archivos, existen dos métodos análogos a los de lectura:

csv.Writer() y csv.DictWriter()

In [33]:
# Ejemplo de DictWriter

destination_filename = os.path.join(os.getcwd(), 'test-csv-writer.csv') # siempre convienen los paths absolutos

with open(destination_filename, 'w') as fp: # el handler para evitar EOF y tener que cerrar
    column_names = data[0].keys() # usamos para el nombre de cada columna las keys
    writer = csv.DictWriter(fp, fieldnames=column_names, delimiter= '\t') # crearemos un archivo separado por tabuladores
    writer.writeheader() # escribimos el header del archivo
    for d in data:
        writer.writerow(d)

El archivo guardado luce asi:

Date	Open	High	Low	Close	Adj Close	Volume

2023-01-03	89.589996	91.050003	88.519997	89.120003	89.120003	28131200

2023-01-04	90.349998	90.650002	87.269997	88.080002	88.080002	34854800

2023-01-05	87.470001	87.570000	85.900002	86.199997	86.199997	27194400

2023-01-06	86.790001	87.690002	84.860001	87.339996	87.339996	41381500

2023-01-09	88.360001	90.050003	87.860001	88.019997	88.019997	29003900

.....

# El módulo datetime

Este módulo es sumamente importante para utilizar un tipo de dato derivado de Python. Los objetos datetime tienen definidas operaciones entre si, pueden ser convertidos fácilmente a formatos legibles por humanos entre otras ventajas

In [65]:
import datetime # utilizamos un alias para no escribir datetime cada vez

In [66]:
#para obtener el timestamp actual
ahora = datetime.datetime.now()
print(ahora, type(ahora)) # es un objeto de la clase datetime.datetime

2024-01-04 02:43:31.502686 <class 'datetime.datetime'>


In [67]:
# se puede extraer la fecha
print(ahora.date())
# o el dia
print(ahora.day) # es un atributo por eso no lleva parentesis
# o la hora
print(ahora.time())

2024-01-04
4
02:43:31.502686


In [69]:
# se puede convertir en string facilmente con el formato que se quiera
ahora_string = ahora.strftime("Es la hora %H:%M:%S del dia %d del mes %m del año %Y")
ahora_string

'Es la hora 02:43:31 del dia 04 del mes 01 del año 2024'

In [70]:
# se pueden crear objetos date y datetime a partir de strings
t_string = '02-04-1995 23:04:35.957'
t = datetime.datetime.strptime(t_string, '%d-%m-%Y %H:%M:%S.%f')
print(t_string , '->', t)

02-04-1995 23:04:35.957 -> 1995-04-02 23:04:35.957000


In [71]:
# Además existen los objetos de diferencias de tiempo (timedelta)
# por ejemplo un intervalo de una hora se genera asi:
td = datetime.timedelta(hours =1)
print(td)
# se puede sumar al objeto anterior una hora de la siguiente manera
print(t, t+ td) # observar que solo cambia de fecha al agregar una hora

1:00:00
1995-04-02 23:04:35.957000 1995-04-03 00:04:35.957000


In [73]:
# como ejemplo vamos a cambiar los strings de la tabla de precios de Google 
# por el formato dia/mes/año

for d in data: # recorremos por filas la lista
    aux = datetime.datetime.strptime(d['Date'],"%Y-%m-%d") # leemos en el formato que esta y creamos un objeto datetime
    d['Date'] = aux.strftime('%d/%m/%Y')
print(data[3])  # el formato ha cambiado

{'Date': '06/01/2023', 'Open': '86.790001', 'High': '87.690002', 'Low': '84.860001', 'Close': '87.339996', 'Adj Close': '87.339996', 'Volume': '41381500'}
